### PR TITLE
Adds Hugging Face integration

### PR DESCRIPTION
**Pull request by mir-am**
# Description of changes made
This PR adds the Hugging Face platform (an alternative to OpenAI) to TestSpark and makes the following changes:
- Adds the Hugging Face option to the LLM setup panel and two models, namely, `Meta-Llama-3-70B-Instruct` & `Meta-Llama-8B-Instruct`
- Adds the HF option to the plugin's settings plus the two mentioned models
- Extends `JunitTestsAssembler` to handle HF requests
- Creates `HuggingFaceRequestsManager` to send inference requests to the HF API

# Why is merge request needed
This PR addresses the issue #167. Also, it would be nice to use some of the latest LLMs on HF like Llama 3 for test case generation.

# Other notes
- Users need to have an [HF API token](https://huggingface.co/inference-api/serverless) similar to OpenAI's.
- For `Meta-Llama-3-70B-Instruct`, a Pro subscription is required on the HF platform to query the model.

# What is missing?
The documentation (`README`) should be updated for using the HF models.

- [x] I have checked that I am merging into the correct branch

### CODE REVIEW

**Comment by arksap2002** [comment]
Thank you @mir-am for adding the Hugging Face in the plugin, this is very useful for us! Could you, please, help me with using this platform. I was trying to apply token from the hugging face profile, which I creted. But, unfortunately, they are not correct, how can I find a correct one?

**Comment by mir-am** [comment]
I can fix the Klint errors in the build logs.

**Comment by arksap2002** [comment]
I will fix them right now, no problem, we have a special script `ktlint.sh` for this.
Huge thanks for implementing this part, it is really useful for the plugin! :)
I will ask the team about the way of getting HF token for usage in our plugin, and after that, I'll review the PR.

**Comment by arksap2002** [comment]
UPD: I can't push in your branch. Could you, please, run `./ktlint.sh` and push the changes :)

**Comment by mir-am** [comment]
> UPD: I can't push in your branch. Could you, please, run `./ktlint.sh` and push the changes :)

I'll take care of it now.

**Comment by arksap2002** [comment]
> > UPD: I can't push in your branch. Could you, please, run `./ktlint.sh` and push the changes :)
> 
> I'll take care of it now.

thank you!

**Comment by mir-am** [comment]
> I will fix them right now, no problem, we have a special script `ktlint.sh` for this. Huge thanks for implementing this part, it is really useful for the plugin! :) I will ask the team about the way of getting HF token for usage in our plugin, and after that, I'll review the PR.

For testing, you can simply sign up to Hugging Face, get an API key, and paste it into the token field for HF in the LLM Setup panel.
Also, please note that you can only use the model, `Meta-Llama-8B-Instruct` with a free subscription on HF. The bigger Llama model needs a PRO subscription.

**Comment by arksap2002** [comment]
> > I will fix them right now, no problem, we have a special script `ktlint.sh` for this. Huge thanks for implementing this part, it is really useful for the plugin! :) I will ask the team about the way of getting HF token for usage in our plugin, and after that, I'll review the PR.
> 
> For testing, you can simply sign up to Hugging Face, get an API key, and paste it into the token field for HF in the LLM Setup panel. Also, please note that you can only use the model, `Meta-Llama-8B-Instruct` with a free subscription on HF. The bigger Llama model needs a PRO subscription.

Do I understand correctly, that I have to get the tokens from the "Access Tokens" part of my HF account? Unfortunately, none of the three types work

**Comment by mir-am** [comment]
> > > I will fix them right now, no problem, we have a special script `ktlint.sh` for this. Huge thanks for implementing this part, it is really useful for the plugin! :) I will ask the team about the way of getting HF token for usage in our plugin, and after that, I'll review the PR.
> > 
> > 
> > For testing, you can simply sign up to Hugging Face, get an API key, and paste it into the token field for HF in the LLM Setup panel. Also, please note that you can only use the model, `Meta-Llama-8B-Instruct` with a free subscription on HF. The bigger Llama model needs a PRO subscription.
> 
> Do I understand correctly, that I have to get the tokens from the "Access Tokens" part of my HF account? Unfortunately, none of the three types work

Yes! You can get an Access Token for your HF account. Mine has a read access and it was working fine. I will test with a free HF account to see if I can reproduce the issue. 
Could you post the error message? When debugging, it should show an error if it can't connect to HF.

**Comment by arksap2002** [comment]
> > > > I will fix them right now, no problem, we have a special script `ktlint.sh` for this. Huge thanks for implementing this part, it is really useful for the plugin! :) I will ask the team about the way of getting HF token for usage in our plugin, and after that, I'll review the PR.
> > > 
> > > 
> > > For testing, you can simply sign up to Hugging Face, get an API key, and paste it into the token field for HF in the LLM Setup panel. Also, please note that you can only use the model, `Meta-Llama-8B-Instruct` with a free subscription on HF. The bigger Llama model needs a PRO subscription.
> > 
> > 
> > Do I understand correctly, that I have to get the tokens from the "Access Tokens" part of my HF account? Unfortunately, none of the three types work
> 
> Yes! You can get an Access Token for your HF account. Mine has a read access and it was working fine. I will test with a free HF account to see if I can reproduce the issue. Could you post the error message? When debugging, it should show an error if it can't connect to HF.

I have this probelm: `The result is invalid or uses unknown commands due to randomness and lack of guarantees from Large Language Model. Please try again`. I also added my screen video. Which class for testing did you use, maybe it's a problem from our side with our parser?

Uploading Screen Recording 2024-05-23 at 3.49.45 PM.mov…

**Comment by mir-am** [comment]
@arksap2002, thanks for sending the video. I also used a simple `Calculator` class, similar to yours, to test the HF models. Could you please share the Example project on the G. Drive for testing?

**Comment by arksap2002** [comment]
> @arksap2002, thanks for sending the video. I also used a simple `Calculator` class, similar to yours, to test the HF models. Could you please share the Example project on the G. Drive for testing?

sure, https://drive.google.com/drive/folders/1vAKHI8RgsS-BgDIWNad5cs7M_V0jHBql?usp=sharing

**Comment by mir-am** [comment]
> > @arksap2002, thanks for sending the video. I also used a simple `Calculator` class, similar to yours, to test the HF models. Could you please share the Example project on the G. Drive for testing?
> 
> sure, https://drive.google.com/drive/folders/1vAKHI8RgsS-BgDIWNad5cs7M_V0jHBql?usp=sharing

Using your example, I've managed to reproduce the issue. 
I just pushed a fix now. Please try again after updating the project.

**Comment by arksap2002** [comment]
@mir-am thank you for adding these fixes. And I am so sorry, this was probably some issue with my token because Pouria created it with his account, just shared it with me, and it works. Please sorry again!

There is an issue, which we found. I can fix it by myself if you want. The situation is when the user doesn't have a PRO account and generates tests with the incorrect model. It is important to show them an error message instead of a plugin's crash.

**Comment by mir-am** [comment]
> @mir-am thank you for adding these fixes. And I am so sorry, this was probably some issue with my token because Pouria created it with his account, just shared it with me, and it works. Please sorry again!
> 
> There is an issue, which we found. I can fix it by myself if you want. The situation is when the user doesn't have a PRO account and generates tests with the incorrect model. It is important to show them an error message instead of a plugin's crash.

No worries! Good to hear that it is working for you.
I will fix the described issue where the user has a free account, and it should show a proper error message.

File: src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt [hunk 1]
```diff
@@ -48,22 +49,33 @@ class HuggingFaceRequestManager(project: Project) : IJRequestManager(project) {
...
-                HttpURLConnection.HTTP_INTERNAL_ERROR -> {
-                    llmErrorManager.errorProcess(
-                        LLMMessagesBundle.get("serverProblems"),
-                        project,
-                        errorMonitor,
-                    )
-                    sendResult = SendResult.OTHER
+                    HttpURLConnection.HTTP_BAD_REQUEST -> {
+                        llmErrorManager.errorProcess(
+                            "The selected model may need an HF PRO subscription to use!",
```
Comment (arksap2002) [inline src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt:68]: could you, please, move this text to the `LLMMessages.properties` and call `LLMMessagesBundle.get(...)` here

**Review by arksap2002 — CHANGES_REQUESTED** [review]

File: src/main/kotlin/org/jetbrains/research/testspark/helpers/LLMHelper.kt [hunk 2]
```diff
@@ -325,4 +339,13 @@ object LLMHelper {
...
         }
     }
+
+    /**
+     * Retrieves the available HuggingFace models.
+     *
+     * @return an array of string representing the available HuggingFace models
+     */
+    private fun getHuggingFaceModels(): Array<String> {
+        return arrayOf("Meta-Llama-3-8B-Instruct", "Meta-Llama-3-70B-Instruct")
```
Comment (Vladislav0Art) [inline src/main/kotlin/org/jetbrains/research/testspark/helpers/LLMHelper.kt:349]: Does HuggingFace provide an API for retrieving available models? 

For instance, to get OpenAI's models, the function `getOpenAIModels` (in this file) requests the corresponding OpenAI API. A similar implementation would be appreciated if the Hugging Face platform provides the respective APIs.

File: src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt (hunk 1 above)
Comment (arksap2002) [reply src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt]: thank you for fixing

**Review by arksap2002 — APPROVED** [review]

File: src/main/kotlin/org/jetbrains/research/testspark/helpers/LLMHelper.kt (hunk 2 above)
Comment (mir-am) [reply src/main/kotlin/org/jetbrains/research/testspark/helpers/LLMHelper.kt:349]: Yes, the HF seems to provide [an API](https://huggingface.co/docs/hub/en/api#get-apimodels) to fetch available models on their platform. However, many models on the HF platform are not necessarily suitable for test case generation. For this PR, I chose two new powerful LLMs that can do well for the said task.

I would suggest gradually adding more HF models in the near future, after testing their performance for test case generation.

File: src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/JUnitTestsAssembler.kt [hunk 3]
```diff
@@ -85,6 +85,30 @@ class JUnitTestsAssembler(
...
+    }
+
+    /**
+     * Removes delimiter lines used for code blocks in LLMs' response
+     */
+    private fun removeDelimiterLines(text: String): String {
+        return text.lines()
+            .filter { line -> line != "```" && line != "```java" }
+            .joinToString("\n")
+    }
```
Comment (Vladislav0Art) [inline src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/JUnitTestsAssembler.kt:110]: Do I understand correctly that this function serves merely as a "hack" to avoid empty triple-tick blocks? 

As I see from the HF request body, oftentimes, the LLM's response contains an empty block of ` ``` `, and only after provides a test suite (the selected model is `Meta-Llama-3-8B-Instruct`).

**Comment by Vladislav0Art** [comment]
The `Meta-Llama-3-8B-Instruct`'s generation success rate is poor. Most of my test generation attempts fail. Is this due to its relatively modest generative capabilities?

If you have a PRO subscription, how do you find the success rate of the 2nd model (is it practically usable)? The `Meta-Llama-3-8B-Instruct` yields a very poor user experience.

File: src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/JUnitTestsAssembler.kt (hunk 3 above)
Comment (mir-am) [reply src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/JUnitTestsAssembler.kt:110]: The function extracts generated test suites from the LLM's response.

**Comment by mir-am** [comment]
> The `Meta-Llama-3-8B-Instruct`'s generation success rate is poor. Most of my test generation attempts fail. Is this due to its relatively modest generative capabilities?
> 
> If you have a PRO subscription, how do you find the success rate of the 2nd model (is it practically usable)? The `Meta-Llama-3-8B-Instruct` yields a very poor user experience.

Thanks for testing the PR.
The model ` Meta-Llama-3-8B-Instruct` is a general-purpose LLM, and its performance for unit test generation can be poor depending on the input code.
Also, the `80B` version of the model performs almost the same for the said task.
I chose these two models to add the HF integration to the plugin. Later, it is possible to add better models after testing their performance on the unit test generation.
For now, you can view the HF integration as an "experimental" feature.

File: src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/JUnitTestsAssembler.kt (hunk 3 above)
Comment (Vladislav0Art) [reply src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/JUnitTestsAssembler.kt:110]: There is an oversight in this implementation. Let me explain by the example test suite generated by `Meta-Llama-3-8B-Instruct`:

Having the following prompt:
````log
<s>[INST] <<SYS>> You are a helpful and honest code and programming assistant. Please, respond concisely and truthfully. <</SYS>> Generate unit tests in Java for Calc.
All tests should be for JUnit 4.
All test cases MUST be inserted into a single test suite.
Name all methods according to the template - [MethodUnderTest][Scenario]Test, and use only English letters.
The source code of class under test is as follows:
```
public class Calc {
    public int sum(int a, int b) {
        return a + b;
    }

    public int multiply(int a, int b) {
        return a * b;
    }
}
```

Generate public classes and public methods. Response only a code with tests between ```, do not provide any other text. [/INST]
````
The LLM responds with the following output:
````log
```

```java
```java
import org.junit.Before;
import org.junit.Test;
import static org.junit.Assert.assertEquals;

public class CalcTest {
    private Calc calc;

    @Before
    public void setup() {
        calc = new Calc();
    }

    @Test
    public void sumTwoPositiveNumbersTest() {
        assertEquals(4, calc.sum(2, 2));
    }

    @Test
    public void sumTwoNegativeNumbersTest() {
        assertEquals(-4, calc.sum(-2, -2));
    }

    @Test
    public void sumPositiveAndNegativeNumberTest() {
        assertEquals(0, calc.sum(2, -2));
    }

    @Test
    public void multiplyTwoPositiveNumbersTest() {
        assertEquals(6, calc.multiply(2, 3));
    }

    @Test
    public void multiplyTwoNegativeNumbersTest() {
        assertEquals(-6, calc.multiply(-2, -3));
    }

    @Test
    public void multiplyPositiveAndNegativeNumberTest() {
        assertEquals(-6, calc.multiply(2, -3));
    }
}
```java``````java```
```java``````java```
```java``````java```
````
The problem arises for the trailing triple-tick blocks at the end: The condition within `filter` of `removeDelimiterLines` is true for them, as the tick blocks appear on the same line (therefore, the `line` is neither \`\`\`-block nor \`\`\`java-block).

### Why it breaks things:

Our parser `org.jetbrains.research.testspark.core.test.parsers.java.JUnitTestSuiteParser` has a method `parseTestSuite` which is responsible for parsing the test suite from the text into a `TestSuiteGeneratedByLLM` DTO structure, and it has the following code:
```java
var rawCode = rawText

if (rawText.contains("```")) {
    rawCode = rawText.split("```")[1]
}
```
This piece of code on the above LLM response will yield the string `"java"`, not the test suite text, therefore the test suite is lost.

Responses alike the above are very common, if not normal, at least to `Meta-Llama-3-8B-Instruct` (maybe for the 2nd model it is different; cannot test without the paid subscription).

### How to fix
These empty triple-tick blocks should be sieved out from the response; it may be done as follows:
```kotlin
private fun removeDelimiterLines(text: String): String {
    return text
        .replace("```java", "```") // making closing tick-blocks be the same as the opening ones
        .split("```") // splitting into potential code chunks; we assume that only one contains the test suite
        .filter { it.isNotBlank() } // remove empty code chunks (assuming all of them are empty but one with the test suite)
        .joinToString { codeChunk -> "```\n$codeChunk\n```" } // embrace the code chunk with the test suite into tick-block, so it can be parsed by `JUnitTestSuiteParser` further
}
```
Perhaps, the proposal might not cover some edge-case responses, but even as is it significantly increases the success rate making the LLM usable.

File: src/main/kotlin/org/jetbrains/research/testspark/helpers/LLMHelper.kt [hunk 4]
```diff
@@ -66,6 +67,9 @@ object LLMHelper {
             if (platformSelector.selectedItem!!.toString() == settingsState.grazieName) {
                 modules = getGrazieModels()
             }
+            if (platformSelector.selectedItem!!.toString() == settingsState.huggingFaceName) {
+                modules = getHuggingFaceModels()
```
Comment (Vladislav0Art) [inline src/main/kotlin/org/jetbrains/research/testspark/helpers/LLMHelper.kt:71]: We likely had a typo in the variable name; it should have been `models`. I created an [issue](https://github.com/JetBrains-Research/TestSpark/issues/268) for this. You may either leave it as is or fix it if you want to.

File: src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestBody.kt [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+)
+
+data class HuggingFaceRequestBody(
+    val messages: List<ChatMessage>,
+    val parameters: Parameters,
+)
+
+fun HuggingFaceRequestBody.toMap(): Map<String, Any> {
+    return mapOf(
+        "inputs" to this.messages.joinToString(separator = "\n") { it.content },
```
Comment (Vladislav0Art) [inline src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestBody.kt:17]: Could you add a small Javadoc comment for this method and a link to the platform docs where the `parameters` are described?

File: src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+import org.jetbrains.research.testspark.tools.llm.generation.JUnitTestsAssembler
+import java.net.HttpURLConnection
+
+/**
+ * A class to manage requests sent to large language models hosted on HuggingFace
+ */
+class HuggingFaceRequestManager(project: Project) : IJRequestManager(project) {
+    private val url = "https://api-inference.huggingface.co/models/meta-llama/"
+    private val systemPrompt = "You are a helpful and honest code and programming assistant." +
+        " Please, respond concisely and truthfully."
```
Comment (Vladislav0Art) [inline src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt:24]: Please, move the string literal into `resources/properties/llm/LLMDefaults.properties`. You may access it as `LLMDefaultsBundle.get("HuggingFaceInitialSystemPrompt")`.

File: src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+    // TODO: The user should be able to change these numbers in the plugin's settings
+    private val topProbability = 0.9
+    private val temperature = 0.9
```
Comment (Vladislav0Art) [inline src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt:28]: Are you planning to introduce the settings options for these constants in this PR?

@arksap2002 may explain how to create a settings page for Hugging Face specifically (as I see it is **not created**) and add the additional variables there.

File: src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+            log.error { "Error in sending request: ${e.message}" }
+        }
+        return sendResult
+    }
+
+    /**
+     * Creates the required prompt for Llama models. For more details see:
+     * https://huggingface.co/blog/llama2#how-to-prompt-llama-2
+     */
+    private fun formatPrompt(systemPrompt: String, userMessage: String): String {
```
Comment (Vladislav0Art) [inline src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt:86]: `systemPrompt` will never change in a runtime; once you move it into the `LLMDefaults.properties` you may access it directly here. 

I suggest naming the method `craftInstructionPrompt` or `createInstructionPrompt`; `formatPrompt` implies a kind of generic text formatting detached from the LLM context.

Comment (Vladislav0Art) [reply src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt:86]: I understood that this prompt modification is only required for the `Llama` family. Since we may integrate other non-Llama models in the future, the implementation will have to be modified. Could you leave a `TODO` comment concerning this aspect?

**Review by Vladislav0Art — CHANGES_REQUESTED** [review]
I left some comments that should improve the robustness of the implementation.

Thank you for contributing such a complex and valuable feature!

**Comment by mir-am** [comment]
> I left some comments that should improve the robustness of the implementation.
> 
> Thank you for contributing such a complex and valuable feature!

Thanks for the thorough code review.
I will address your comments.

File: src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/JUnitTestsAssembler.kt (hunk 3 above)
Comment (mir-am) [reply src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/JUnitTestsAssembler.kt]: I've implemented a similar fix in https://github.com/JetBrains-Research/TestSpark/pull/212/commits/83bdd2be0577de49cd277f18755be7b6eb9a2fa7 to make the LLM response compatible with the parser.

[the remark this answers is not in the document]
Comment (mir-am) [reply src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/JUnitTestsAssembler.kt]: I agree that this is a valid concern. I've moved the post-processing logic to `HuggingFaceRequestManager` in https://github.com/JetBrains-Research/TestSpark/pull/212/commits/4bce7036cbf6dac6a743aaa7332ccb5a9b13f7c3 and left a TODO to create a generic `consume(String)` later.

File: src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestBody.kt (hunk 5 above)
Comment (mir-am) [reply src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestBody.kt:21]: Addressed in https://github.com/JetBrains-Research/TestSpark/pull/212/commits/e93a3ed998c961d36c5675f1cff1e41b633f74a2

File: src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt (hunk 7 above)
Comment (mir-am) [reply src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt:28]: Addressed in https://github.com/JetBrains-Research/TestSpark/pull/212/commits/0b00e8db3c3c474e63c765b3864c2c7991b5fa4b

File: src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt (hunk 6 above)
Comment (mir-am) [reply src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt]: Addressed in https://github.com/JetBrains-Research/TestSpark/pull/212/commits/0b00e8db3c3c474e63c765b3864c2c7991b5fa4b

File: src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt (hunk 7 above)
Comment (mir-am) [reply src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt:28]: I would like to create another PR to allow users to set those variables in the plugin's settings.

File: src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt (hunk 8 above)
Comment (mir-am) [reply src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt]: Added a TODO in https://github.com/JetBrains-Research/TestSpark/pull/212/commits/0b00e8db3c3c474e63c765b3864c2c7991b5fa4b.

File: src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/JUnitTestsAssembler.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (Vladislav0Art) [reply src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/JUnitTestsAssembler.kt]: > left a TODO to create a generic consume(String) later.

It already exists: `JUnitTestsAssembler` implements the `TestsAssembler` interface that has the `consume(String)` method ([here](https://github.com/JetBrains-Research/TestSpark/blob/e6adea3171f96719d5bb7853e782d9ba4e4c0ac9/src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/JUnitTestsAssembler.kt#L46) is the `JUnitTestsAssembler`'s override). You may use it instead of `consumeHFRequest`.

[the remark this answers is not in the document]
Comment (mir-am) [reply src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/JUnitTestsAssembler.kt]: I've used `consume(String)` now (https://github.com/JetBrains-Research/TestSpark/pull/212/commits/269353925813e2542156d1031462f56ac4f7b5d9)

File: src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt (hunk 7 above)
Comment (Vladislav0Art) [reply src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt:28]: Agree. Once this PR is merged, I will create several issues reflecting the settings page and the TODOs.

I also recommend adding those hard-coded parameters in the `HuggingFaceRequestBody.toMap()` function to the settings page.

File: src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+                it.write(GsonBuilder().disableHtmlEscaping().create().toJson(llmRequestBody))
+                when (val responseCode = (it.connection as HttpURLConnection).responseCode) {
+                    HttpURLConnection.HTTP_OK -> {
+                        val text = it.reader.readLine()
+                        val generatedTestCases = extractLLMGeneratedCode(
+                            JsonParser.parseString(text).asJsonArray[0]
+                                .asJsonObject["generated_text"].asString.trim(),
+                        )
+                        (testsAssembler as JUnitTestsAssembler).consume(generatedTestCases)
+                    }
```
Comment (Vladislav0Art) [inline src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt:67]: Since we use the method coming from the interface, it can be simply:
```java
testsAssembler.consume(generatedTestCases)
```
No cast is needed; it is even required to remove the cast since other languages, such as Kotlin, will be coming soon.

File: src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+        // TODO: This is Llama-specific and should support other LLMs hosted on HF too.
+        return "<s>[INST] <<SYS>> ${LLMDefaultsBundle.get("huggingFaceInitialSystemPrompt")} <</SYS>> $userMessage [/INST]"
+    }
+
+    /**
+     * Extracts code blocks in LLMs' response.
+     * Also, it handles the cases where the LLM-generated code does not end with ```
+     */
+    private fun extractLLMGeneratedCode(text: String): String {
+        val modifiedText = text.replace("```java", "```").replace("````", "```")
```
Comment (Vladislav0Art) [inline src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt:108]: Also, one point: the `java` is hard-coded here. We need to figure out the language in which the test generation was requested dynamically. But let's have another PR for this, as it is not that trivial and may require intrusive changes; a `TODO` will be enough for now here.

Comment (Vladislav0Art) [reply src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt:108]: Why is `.replace("````", "```")` needed? It seems not to affect anything. 

I've never seen the LLM responding with four backtick blocks, although it may merge several triple backtick blocks, resulting in the `"``````"` block.

**Review by Vladislav0Art — COMMENTED** [review]
I think that is the last iteration. There are some minor points from my side. All the rest is good. Thank you!

**Comment by Vladislav0Art** [comment]
There is a small conflict in the `LLMHelper.kt` file; to resolve it, you only need to change the variable name from `modules` to `models`.

File: src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt (hunk 9 above)
Comment (mir-am) [reply src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt]: Addressed in https://github.com/JetBrains-Research/TestSpark/pull/212/commits/e218dfa573d4bd55d4ede21efdc17edcac461852.

File: src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt (hunk 10 above)
Comment (mir-am) [reply src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt:109]: I've seen it twice with  ` ```` `. It can be removed, though.

**Comment by Vladislav0Art** [comment]
@mir-am Hi! 
There are some conflicts again. I resolved them in the [mir-am-hf-integration](https://github.com/JetBrains-Research/TestSpark/tree/mir-am-hf-integration) branch. You may incorporate them. The main changes:

1. In `LLMHelpers`: change the `modules` to `models` variable name + add hugging-face-related code.
2. The `SettingsArguments` class in `SettingsArguments.kt` has changed its name to `LlmSettingsArguments` (the file name is also changed to `LlmSettingsArguments.kt`).

**Comment by mir-am** [comment]
> @mir-am Hi! There are some conflicts again. I resolved them in the [mir-am-hf-integration](https://github.com/JetBrains-Research/TestSpark/tree/mir-am-hf-integration) branch. You may incorporate them. The main changes:
> 
> 1. In `LLMHelpers`: change the `modules` to `models` variable name + add hugging-face-related code.
> 2. The `SettingsArguments` class in `SettingsArguments.kt` has changed its name to `LlmSettingsArguments` (the file name is also changed to `LlmSettingsArguments.kt`).

Thanks! I've resolved the conflicts using your suggestion.

**Comment by Vladislav0Art** [comment]
I see 51 files modified now.. That is weird. I expected very few changes compared to the version before the conflict resolution.

For some reason, GitHub diff shows the difference in changes that should have been in the `development` branch already. I will check whether everything is okay with, but it may require to revert the changes to [c9d00c8](https://github.com/JetBrains-Research/TestSpark/pull/212/commits/c9d00c8c02b29f0e2ff26fe0a00fc280d30d7f09) commit (i.e., your last commit before the merge of `mir-am-hf-integration` branch, and then resolve the conflict again but manually.

**Comment by mir-am** [comment]
> I see 51 files modified now.. That is weird. I expected very few changes compared to the version before the conflict resolution.
> 
> For some reason, GitHub diff shows the difference in changes that should have been in the `development` branch already. I will check whether everything is okay with, but it may require to revert the changes to [c9d00c8](https://github.com/JetBrains-Research/TestSpark/pull/212/commits/c9d00c8c02b29f0e2ff26fe0a00fc280d30d7f09) commit (i.e., your last commit before the merge of `mir-am-hf-integration` branch, and then resolve the conflict again but manually.

Thanks for the help.
Let me know if I need to resolve the conflicts manually.

**Comment by Vladislav0Art** [comment]
Okay, it works fine. 

**Thank you for your valuable contribution! 👍🏻 🔥**
